api/v1alpha1/contour_types.go

```diff
@@ -91,6 +98,155 @@ type NamespaceSpec struct {
 	RemoveOnDeletion bool `json:"removeOnDeletion,omitempty"`
 }
 
+// NetworkPublishing defines the schema for publishing Contour to a network.
+type NetworkPublishing struct {
+	// Envoy provides the schema for publishing the network endpoints of Envoy.
+	//
+	// If unset, defaults to:
+	//   type: LoadBalancerService
+	//   containerPorts:
+	//   - name: http
+	//     portNumber: 8080
+	//   - name: https
+	//     portNumber: 8443
+	//
+	// +kubebuilder:default={type: LoadBalancerService, containerPorts: {{name: http, portNumber: 8080}, {name: https, portNumber: 8443}}}
+	Envoy EnvoyNetworkPublishing `json:"envoy,omitempty"`
+}
+
+// EnvoyNetworkPublishing defines the schema to publish Envoy to a network.
+// +union
+type EnvoyNetworkPublishing struct {
+	// Type is the type of publishing strategy to use. Valid values are:
+	//
+	// * LoadBalancerService
+	//
+	// In this configuration, network endpoints for Envoy use container networking.
+	// A Kubernetes LoadBalancer Service is created to publish Envoy network
+	// endpoints. The Service uses port 80 to publish Envoy's HTTP network endpoint
+	// and port 443 to publish Envoy's HTTPS network endpoint.
+	//
+	// See: https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer
+	//
+	// * NodePortService
+	//
+	// Publishes Envoy network endpoints using a Kubernetes NodePort Service.
+	//
+	// In this configuration, Envoy network endpoints use container networking. A Kubernetes
+	// NodePort Service is created to publish the network endpoints.
+	//
+	// See: https://kubernetes.io/docs/concepts/services-networking/service/#nodeport
+	//
+	// +unionDiscriminator
+	// +kubebuilder:default=LoadBalancerService
+	Type NetworkPublishingType `json:"type,omitempty"`
+
+	// loadBalancer holds parameters for the load balancer. Present only if type is
+	// LoadBalancerService.
+	//
+	// If unspecified, defaults to an external Classic AWS ELB.
+	//
+	LoadBalancer LoadBalancerStrategy `json:"loadBalancer,omitempty"`
+
+	// ContainerPorts is a list of container ports to expose from the Envoy container(s).
```

Comment:
@youngnick @stevesloka @skriss do you agree that at least 2 `ContainerPorts` must be defined to support Envoy's insecure and secure services? I'm unsure if it's possible or if a use-case exists to run the insecure/secure services on the same container port. Do you agree that 6 container ports is an appropriate maximum?

Reply:
I think those are totally acceptable numbers for now. We can change this later.